core/src/main/scala/com/salesforce/op/ModelInsights.scala

```diff
@@ -443,43 +444,69 @@ case object ModelInsights {
-    val model = models.lastOption
+      case s: SelectedCombinerModel => s
+    }
+    val model = models.flatMap{
+      case s: SelectedCombinerModel if s.strategy == CombinationStrategy.Best =>
+        val originF = if (s.weight1 > 0.5) s.getInputFeature[Prediction](1) else s.getInputFeature[Prediction](2)
+        models.find( m => originF.exists(_.originStage.uid == m.uid) )
+      case s => Option(s)
+    }.lastOption
+
     log.info(
       s"Found ${models.length} models will " +
         s"${model.map("use results from the last model:" + _.uid + "to").getOrElse("not")}" +
         s" to fill in model insights"
     )
 
-    val label = model.map(_.getInputFeature[RealNN](0)).orElse(sanityChecker.map(_.getInputFeature[RealNN](0))).flatten
+    val modelInputStages: Set[String] = model.map { m =>
+      val stages = m.getInputFeatures().map(_.parentStages().toOption.map(_.keySet.map(_.uid)))
+      val uid = stages.collect{ case Some(uids) => uids }
+      uid.fold(Set.empty)(_ + _)
+    }.getOrElse(Set.empty)
+
+    val sanityCheckers = stages.collect { case s: SanityCheckerModel => s }
+    val sanityCheckersForModel = sanityCheckers.filter(s => modelInputStages.contains(s.uid) &&
+      model.exists(_.getInputFeature[RealNN](0) == s.getInputFeature[RealNN](0))).toSeq
+
+    val sanityChecker = if (sanityCheckersForModel.nonEmpty) sanityCheckersForModel else sanityCheckers.lastOption.toSeq
+    val checkerSummary = if (sanityChecker.nonEmpty) {
+      Option(SanityCheckerSummary.flatten(
+        sanityChecker.map(s => SanityCheckerSummary.fromMetadata(s.getMetadata().getSummaryMetadata()))
+      ))
+    } else None
+
+    log.info(
+      s"Found ${sanityCheckers.length} sanity checkers" +
+        s"${sanityChecker.map("will preferentially use results from checkers in model path:" + _.uid +
+          " to fill in model insights")}"
+    )
+
+
+    val label = model.map(_.getInputFeature[RealNN](0))
+      .orElse(sanityChecker.lastOption.map(_.getInputFeature[RealNN](0))).flatten
     log.info(s"Found ${label.map(_.name + " as label").getOrElse("no label")} to fill in model insights")
 
     // Recover the vector metadata
     val vectorInput: Option[OpVectorMetadata] = {
       def makeMeta(s: => OpPipelineStageParams) = Try(OpVectorMetadata(s.getInputSchema().last)).toOption
 
-      sanityChecker
-        // first try out to get vector metadata from sanity checker
-        .flatMap(s => makeMeta(s.parent.asInstanceOf[SanityChecker]).orElse(makeMeta(s)))
-        // fall back to model selector stage metadata
-        .orElse(model.flatMap(m => makeMeta(m.parent.asInstanceOf[ModelSelector[_, _]])))
-        // finally try to get it from the last vector stage
-        .orElse(
-        stages.filter(_.getOutput().isSubtypeOf[OPVector]).lastOption
-          .map(v => OpVectorMetadata(v.getOutputFeatureName, v.getMetadata()))
-      )
+      if (sanityChecker.nonEmpty) { // first try out to get vector metadata from sanity checker
```

Comment:
I dont think it makes sense to have the flatten function return an option (also I am using an already defined function which has other uses here). I could make a reduce function to wrap the flatten and do a check but that makes the calculation less efficient...though I suppose these are small enough it probably doesnt matter. @tovbinm what is your vote?

Reply:
The way `SanityCheckerSummary.flatten` is defined i.e. `def flatten(checkers: Seq[SanityCheckerSummary]): SanityCheckerSummary` - it should be renamed to `reduce`.

Reply:
Well I can rename it, then it is inconsistent with the OpVectorMetadata functions naming which does exactly the same thing. Also @Juantbox wants a change in the functionality so that it returns a None rather than an Empty summary, which I can easily do for the Sanity Checker but would change the usage of the OpVectorMetadata in a couple of places

Reply:
My 2c: Empty summary (with all `0.0` values I presume) makes more sense to me. Option would create unnecessary clutter.